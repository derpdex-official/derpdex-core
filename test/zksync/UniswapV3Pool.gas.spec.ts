// import { ethers, waffle } from 'hardhat'
// import { Wallet } from 'ethers'
import { MockTimeUniswapV3Pool } from '../../typechain/MockTimeUniswapV3Pool'
import { expect } from './shared/expect'

import { poolFixture } from './shared/fixtures'
// import snapshotGasCost from './shared/snapshotGasCost' // snapshotGasCost affect expect result in zksync localhost test

import {
  expandTo18Decimals,
  FeeAmount,
  getMinTick,
  encodePriceSqrt,
  TICK_SPACINGS,
  createPoolFunctions,
  SwapFunction,
  MintFunction,
  getMaxTick,
  MaxUint128,
  SwapToPriceFunction,
  MAX_SQRT_RATIO,
  MIN_SQRT_RATIO,
} from './shared/utilities'

import { Provider, Wallet } from 'zksync-web3'
import { getSigners } from './shared/zkUtils'
let tx

// const createFixtureLoader = waffle.createFixtureLoader

describe('UniswapV3Pool gas tests', () => {
  let wallet: Wallet, other: Wallet

  // let loadFixture: ReturnType<typeof createFixtureLoader>

  before('create fixture loader', async () => {
    // ;[wallet, other] = await (ethers as any).getSigners()
    ;[wallet, other] = await getSigners()
    // loadFixture = createFixtureLoader([wallet, other])
  })

  for (const feeProtocol of [0, 6]) {
    describe(feeProtocol > 0 ? 'fee is on' : 'fee is off', () => {
      const startingPrice = encodePriceSqrt(100001, 100000)
      const startingTick = 0
      const feeAmount = FeeAmount.MEDIUM
      const tickSpacing = TICK_SPACINGS[feeAmount]
      const minTick = getMinTick(tickSpacing)
      const maxTick = getMaxTick(tickSpacing)

      const gasTestFixture = async ([wallet]: Wallet[]) => {
        // const fix = await poolFixture([wallet], waffle.provider)
        const fix = await poolFixture([wallet], Provider.getDefaultProvider())

        const pool = await fix.createPool(feeAmount, tickSpacing)

        const { swapExact0For1, swapToHigherPrice, mint, swapToLowerPrice } = await createPoolFunctions({
          swapTarget: fix.swapTargetCallee,
          token0: fix.token0,
          token1: fix.token1,
          pool,
        })

        tx = await pool.initialize(encodePriceSqrt(1, 1))
        await tx.wait()
        tx = await pool.setFeeProtocol(feeProtocol, feeProtocol)
        await tx.wait()
        tx = await pool.increaseObservationCardinalityNext(4)
        await tx.wait()
        tx = await pool.advanceTime(1)
        await tx.wait()
        tx = await mint(wallet.address, minTick, maxTick, expandTo18Decimals(2))
        await tx.wait()

        tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
        await tx.wait()
        tx = await pool.advanceTime(1)
        await tx.wait()
        tx = await swapToHigherPrice(startingPrice, wallet.address)
        await tx.wait()
        tx = await pool.advanceTime(1)
        await tx.wait()
        expect((await pool.slot0()).tick).to.eq(startingTick)
        expect((await pool.slot0()).sqrtPriceX96).to.eq(startingPrice)

        return { pool, swapExact0For1, mint, swapToHigherPrice, swapToLowerPrice }
      }

      let swapExact0For1: SwapFunction
      let swapToHigherPrice: SwapToPriceFunction
      let swapToLowerPrice: SwapToPriceFunction
      let pool: MockTimeUniswapV3Pool
      let mint: MintFunction

      beforeEach('load the fixture', async () => {
        // ; ({ swapExact0For1, pool, mint, swapToHigherPrice, swapToLowerPrice } = await loadFixture(gasTestFixture))
        ; ({ swapExact0For1, pool, mint, swapToHigherPrice, swapToLowerPrice } = await gasTestFixture([wallet]))
      })

      describe('#swapExact0For1', () => {
        it('first swap in block with no tick movement', async () => {
          // await snapshotGasCost(swapExact0For1(2000, wallet.address))
          tx = await swapExact0For1(2000, wallet.address)
          await tx.wait()
          expect((await pool.slot0()).sqrtPriceX96).to.not.eq(startingPrice)
          expect((await pool.slot0()).tick).to.eq(startingTick)
        })

        it('first swap in block moves tick, no initialized crossings', async () => {
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick - 1)
        })

        it('second swap in block with no tick movement', async () => {
          tx = await swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick - 1)
          // await snapshotGasCost(swapExact0For1(2000, wallet.address))
          tx = await swapExact0For1(2000, wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick - 1)
        })

        it('second swap in block moves tick, no initialized crossings', async () => {
          tx = await swapExact0For1(1000, wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick)
          tx = await swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick - 1)
        })

        it('first swap in block, large swap, no initialized crossings', async () => {
          tx = await swapExact0For1(expandTo18Decimals(10), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(-35787)
        })

        it('first swap in block, large swap crossing several initialized ticks', async () => {
          tx = await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          tx = await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          )
          await tx.wait()
          expect((await pool.slot0()).tick).to.eq(startingTick)
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(startingTick - 4 * tickSpacing) // we crossed the last tick
        })

        it('first swap in block, large swap crossing a single initialized tick', async () => {
          tx = await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(startingTick - 2 * tickSpacing) // we crossed the last tick
        })

        it('second swap in block, large swap crossing several initialized ticks', async () => {
          tx = await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          tx = await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
            )
          await tx.wait()
          tx = await swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address)
          await tx.wait()
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(startingTick - 4 * tickSpacing)
        })

        it('second swap in block, large swap crossing a single initialized tick', async () => {
          tx = await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          tx = await swapExact0For1(expandTo18Decimals(1).div(10000), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.gt(startingTick - 2 * tickSpacing) // we didn't cross the initialized tick
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(startingTick - 2 * tickSpacing) // we crossed the last tick
        })

        it('large swap crossing several initialized ticks after some time passes', async () => {
          tx = await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          tx = await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
            )
          await tx.wait()
          tx = await swapExact0For1(2, wallet.address)
          await tx.wait()
          tx = await pool.advanceTime(1)
          await tx.wait()
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(startingTick - 4 * tickSpacing)
        })

        it('large swap crossing several initialized ticks second time after some time passes', async () => {
          tx = await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
          await tx.wait()
          tx = await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          )
          await tx.wait()
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          tx = await swapToHigherPrice(startingPrice, wallet.address)
          await tx.wait()
          tx = await pool.advanceTime(1)
          await tx.wait()
          // await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
          tx = await swapExact0For1(expandTo18Decimals(1), wallet.address)
          await tx.wait()
          expect((await pool.slot0()).tick).to.be.lt(tickSpacing * -4)
        })
      })

      describe('#mint', () => {
        for (const { description, tickLower, tickUpper } of [
          {
            description: 'around current price',
            tickLower: startingTick - tickSpacing,
            tickUpper: startingTick + tickSpacing,
          },
          {
            description: 'below current price',
            tickLower: startingTick - 2 * tickSpacing,
            tickUpper: startingTick - tickSpacing,
          },
          {
            description: 'above current price',
            tickLower: startingTick + tickSpacing,
            tickUpper: startingTick + 2 * tickSpacing,
          },
        ]) {
          describe(description, () => {
            it('new position mint first in range', async () => {
              // await snapshotGasCost(mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
            it('add to position existing', async () => {
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
              // await snapshotGasCost(mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
            it('second position in same range', async () => {
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
              // await snapshotGasCost(mint(other.address, tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await mint(other.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
            it('add to position after some time passes', async () => {
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
              tx = await pool.advanceTime(1)
              await tx.wait()
              // await snapshotGasCost(mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
          })
        }
      })

      describe('#burn', () => {
        for (const { description, tickLower, tickUpper } of [
          {
            description: 'around current price',
            tickLower: startingTick - tickSpacing,
            tickUpper: startingTick + tickSpacing,
          },
          {
            description: 'below current price',
            tickLower: startingTick - 2 * tickSpacing,
            tickUpper: startingTick - tickSpacing,
          },
          {
            description: 'above current price',
            tickLower: startingTick + tickSpacing,
            tickUpper: startingTick + 2 * tickSpacing,
          },
        ]) {
          describe(description, () => {
            const liquidityAmount = expandTo18Decimals(1)
            beforeEach('mint a position', async () => {
              tx = await mint(wallet.address, tickLower, tickUpper, liquidityAmount)
              await tx.wait()
            })

            it('burn when only position using ticks', async () => {
              // await snapshotGasCost(pool.burn(tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await pool.burn(tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
            it('partial position burn', async () => {
              // await snapshotGasCost(pool.burn(tickLower, tickUpper, expandTo18Decimals(1).div(2)))
              tx = await pool.burn(tickLower, tickUpper, expandTo18Decimals(1).div(2))
              await tx.wait()
            })
            it('entire position burn but other positions are using the ticks', async () => {
              tx = await mint(other.address, tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
              // await snapshotGasCost(pool.burn(tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await pool.burn(tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
            it('burn entire position after some time passes', async () => {
              tx = await pool.advanceTime(1)
              await tx.wait()
              // await snapshotGasCost(pool.burn(tickLower, tickUpper, expandTo18Decimals(1)))
              tx = await pool.burn(tickLower, tickUpper, expandTo18Decimals(1))
              await tx.wait()
            })
          })
        }
      })

      describe('#poke', () => {
        const tickLower = startingTick - tickSpacing
        const tickUpper = startingTick + tickSpacing

        it('best case', async () => {
          tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
          await tx.wait()
          tx = await swapExact0For1(expandTo18Decimals(1).div(100), wallet.address)
          await tx.wait()
          tx = await pool.burn(tickLower, tickUpper, 0)
          await tx.wait()
          tx = await swapExact0For1(expandTo18Decimals(1).div(100), wallet.address)
          await tx.wait()
          // await snapshotGasCost(pool.burn(tickLower, tickUpper, 0))
          tx = await pool.burn(tickLower, tickUpper, 0)
          await tx.wait()
        })
      })

      describe('#collect', () => {
        const tickLower = startingTick - tickSpacing
        const tickUpper = startingTick + tickSpacing

        it('close to worst case', async () => {
          tx = await mint(wallet.address, tickLower, tickUpper, expandTo18Decimals(1))
          await tx.wait()
          await swapExact0For1(expandTo18Decimals(1).div(100), wallet.address)
          tx = await pool.burn(tickLower, tickUpper, 0) // poke to accumulate fees
          await tx.wait()
          // await snapshotGasCost(pool.collect(wallet.address, tickLower, tickUpper, MaxUint128, MaxUint128))
          tx = await pool.collect(wallet.address, tickLower, tickUpper, MaxUint128, MaxUint128)
          await tx.wait()
        })
      })

      describe('#increaseObservationCardinalityNext', () => {
        it('grow by 1 slot', async () => {
          // await snapshotGasCost(pool.increaseObservationCardinalityNext(5))
          tx = await pool.increaseObservationCardinalityNext(5)
          await tx.wait()
        })
        it('no op', async () => {
          // await snapshotGasCost(pool.increaseObservationCardinalityNext(3))
          tx = await pool.increaseObservationCardinalityNext(3)
          await tx.wait()
        })
      })

      describe('#snapshotCumulativesInside', () => {
        it('tick inside', async () => {
          // await snapshotGasCost(pool.estimateGas.snapshotCumulativesInside(minTick, maxTick))
          await pool.estimateGas.snapshotCumulativesInside(minTick, maxTick)
        })
        it('tick above', async () => {
          tx = await swapToHigherPrice(MAX_SQRT_RATIO.sub(1), wallet.address)
          await tx.wait()
          // await snapshotGasCost(pool.estimateGas.snapshotCumulativesInside(minTick, maxTick))
          pool.estimateGas.snapshotCumulativesInside(minTick, maxTick)
        })
        it('tick below', async () => {
          tx = await swapToLowerPrice(MIN_SQRT_RATIO.add(1), wallet.address)
          await tx.wait()
          // await snapshotGasCost(pool.estimateGas.snapshotCumulativesInside(minTick, maxTick))
          pool.estimateGas.snapshotCumulativesInside(minTick, maxTick)
        })
      })
    })
  }
})