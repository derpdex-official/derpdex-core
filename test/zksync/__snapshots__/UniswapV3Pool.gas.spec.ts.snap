// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniswapV3Pool gas tests fee is off #burn above current price entire position burn but other positions are using the ticks 1`] = `186049`;

exports[`UniswapV3Pool gas tests fee is off #burn around current price burn entire position after some time passes 1`] = `151466`;

exports[`UniswapV3Pool gas tests fee is off #burn around current price burn when only position using ticks 1`] = `147185`;

exports[`UniswapV3Pool gas tests fee is off #burn around current price entire position burn but other positions are using the ticks 1`] = `134535`;

exports[`UniswapV3Pool gas tests fee is off #burn around current price partial position burn 1`] = `190187`;

exports[`UniswapV3Pool gas tests fee is off #burn below current price entire position burn but other positions are using the ticks 1`] = `185923`;

exports[`UniswapV3Pool gas tests fee is off #collect close to worst case 1`] = `166112`;

exports[`UniswapV3Pool gas tests fee is off #increaseObservationCardinalityNext grow by 1 slot 1`] = `152291`;

exports[`UniswapV3Pool gas tests fee is off #increaseObservationCardinalityNext no op 1`] = `145029`;

exports[`UniswapV3Pool gas tests fee is off #mint above current price add to position after some time passes 1`] = `234196`;

exports[`UniswapV3Pool gas tests fee is off #mint above current price add to position existing 1`] = `202828`;

exports[`UniswapV3Pool gas tests fee is off #mint above current price new position mint first in range 1`] = `260153`;

exports[`UniswapV3Pool gas tests fee is off #mint above current price second position in same range 1`] = `204133`;

exports[`UniswapV3Pool gas tests fee is off #mint around current price add to position after some time passes 1`] = `286310`;

exports[`UniswapV3Pool gas tests fee is off #mint around current price add to position existing 1`] = `279514`;

exports[`UniswapV3Pool gas tests fee is off #mint around current price new position mint first in range 1`] = `323737`;

exports[`UniswapV3Pool gas tests fee is off #mint around current price second position in same range 1`] = `281158`;

exports[`UniswapV3Pool gas tests fee is off #mint below current price add to position after some time passes 1`] = `234202`;

exports[`UniswapV3Pool gas tests fee is off #mint below current price add to position existing 1`] = `234202`;

exports[`UniswapV3Pool gas tests fee is off #mint below current price new position mint first in range 1`] = `276136`;

exports[`UniswapV3Pool gas tests fee is off #mint below current price second position in same range 1`] = `235507`;

exports[`UniswapV3Pool gas tests fee is off #poke best case 1`] = `109438`;

exports[`UniswapV3Pool gas tests fee is off #snapshotCumulativesInside tick above 1`] = `178550`;

exports[`UniswapV3Pool gas tests fee is off #snapshotCumulativesInside tick below 1`] = `178550`;

exports[`UniswapV3Pool gas tests fee is off #snapshotCumulativesInside tick inside 1`] = `178550`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `228523`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 first swap in block with no tick movement 1`] = `208926`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `258376`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `332307`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `293177`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `332307`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `365851`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `195455`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `238798`;

exports[`UniswapV3Pool gas tests fee is off #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `279643`;

exports[`UniswapV3Pool gas tests fee is on #burn around current price burn entire position after some time passes 1`] = `151466`;

exports[`UniswapV3Pool gas tests fee is on #burn around current price entire position burn but other positions are using the ticks 1`] = `134535`;

exports[`UniswapV3Pool gas tests fee is on #collect close to worst case 1`] = `114812`;

exports[`UniswapV3Pool gas tests fee is on #mint above current price add to position after some time passes 1`] = `234196`;

exports[`UniswapV3Pool gas tests fee is on #mint above current price add to position existing 1`] = `234196`;

exports[`UniswapV3Pool gas tests fee is on #mint above current price new position mint first in range 1`] = `260153`;

exports[`UniswapV3Pool gas tests fee is on #mint above current price second position in same range 1`] = `204133`;

exports[`UniswapV3Pool gas tests fee is on #mint around current price add to position after some time passes 1`] = `286310`;

exports[`UniswapV3Pool gas tests fee is on #mint around current price add to position existing 1`] = `279853`;

exports[`UniswapV3Pool gas tests fee is on #mint around current price new position mint first in range 1`] = `289854`;

exports[`UniswapV3Pool gas tests fee is on #mint around current price second position in same range 1`] = `281158`;

exports[`UniswapV3Pool gas tests fee is on #mint below current price add to position after some time passes 1`] = `200658`;

exports[`UniswapV3Pool gas tests fee is on #mint below current price add to position existing 1`] = `234202`;

exports[`UniswapV3Pool gas tests fee is on #mint below current price new position mint first in range 1`] = `276136`;

exports[`UniswapV3Pool gas tests fee is on #mint below current price second position in same range 1`] = `201963`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `264842`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 first swap in block with no tick movement 1`] = `245175`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `261077`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `335362`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `262548`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `335362`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `368906`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `198230`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `241573`;

exports[`UniswapV3Pool gas tests fee is on #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `349240`;
